src/ngraph/runtime/cpu/op/max_pool_with_indices.cpp

```diff
@@ -0,0 +1,350 @@
+/*******************************************************************************
+* Copyright 2017-2018 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "ngraph/runtime/cpu/op/max_pool_with_indices.hpp"
+#include "ngraph/function.hpp"
+#include "ngraph/op/add.hpp"
+#include "ngraph/op/constant.hpp"
+#include "ngraph/op/greater.hpp"
+#include "ngraph/op/select_and_scatter.hpp"
+#include "ngraph/util.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+op::MaxPoolWithIndices::MaxPoolWithIndices(const shared_ptr<Node>& arg,
+                                           const Shape& window_shape,
+                                           const Strides& window_movement_strides,
+                                           const Shape& padding_below,
+                                           const Shape& padding_above)
+    : RequiresTensorViewArgs("MaxPoolWithIndices", {arg})
+    , m_window_shape(window_shape)
+    , m_window_movement_strides(window_movement_strides)
+    , m_padding_below(padding_below)
+    , m_padding_above(padding_above)
+{
+    auto& arg_shape = get_input_shape(0);
+
+    //
+    // Make sure arg: NCDi for some Di of rank>0, N != 0, C != 0.
+    //
+    if (arg_shape.size() < 3)
```

Comment:
Can we do something similar to ConvolutionRelu. Create a Maxpool op and that would do all this checking and use that to create a MaxPoolWithIndices.

Reply:
This is actually a **preferred** way as per @aprocter and @diyessi . We kinda cut a few corners for `ConvolutionRelu`. I'd rather change `ConvolutionRelu` to include more checks.